zebedee/client.go

```diff
@@ -0,0 +1,302 @@
+	}
+
+	dlp := new(models.DatasetLandingPage)
+	if err = json.Unmarshal(b, &dlp); err != nil {
+		return StaticDatasetLandingPage{}, err
+	}
+
+	related := [][]model.Related{
+		dlp.RelatedDatasets,
+		dlp.RelatedDocuments,
+		dlp.RelatedMethodology,
+		dlp.RelatedMethodologyArticle,
+	}
+
+	//Concurrently resolve any URIs where we need more data from another page
+	var wg sync.WaitGroup
+	sem := make(chan int, 10)
+
+	for _, element := range related {
+		for i, e := range element {
+			sem <- 1
+			wg.Add(1)
+			go func(i int, e model.Related, element []model.Related) {
+				defer func() {
+					<-sem
+					wg.Done()
+				}()
+				element[i].Title = c.getPageTitle(e.URI)
+			}(i, e, element)
+		}
+	}
+	wg.Wait()
+
+	return c.populateStaticLandingPageModel(*dlp)
+}
+
+func (c ZebedeeClient) SetAccessToken(token string) {
+	log.Debug("adding access token to client", log.Data{"token": token})
+	c.accessToken = token
+}
+
+func (c ZebedeeClient) get(path string) ([]byte, error) {
+	req, err := http.NewRequest("GET", c.zebedeeURL+path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(c.accessToken) > 0 {
+		req.Header.Set("X-Florence-Token", c.accessToken)
+	}
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, ErrInvalidZebedeeResponse{errors.New("unexpected response from zebedee"), zebedeeRequestErrorData{
+			"zebedeeURI":         req.URL.Path,
+			"expectedStatusCode": 200,
+			"actualStatusCode":   resp.StatusCode,
+		}}
+	}
+
+	return ioutil.ReadAll(resp.Body)
+}
+
+func (c ZebedeeClient) getBreadcrumb(uri string) ([]model.TaxonomyNode, error) {
+	b, err := c.get("/parents?uri=" + uri)
+	if err != nil {
+		return nil, err
+	}
+
+	var parentsJSON []models.TaxonomyNode
+	if err = json.Unmarshal(b, &parentsJSON); err != nil {
+		return nil, err
+	}
+
+	var parents []model.TaxonomyNode
+	for _, value := range parentsJSON {
+		parents = append(parents, model.TaxonomyNode{
+			Title: value.Description.Title,
+			URI:   value.URI,
+		})
+	}
+
+	return parents, nil
+}
+
+func (c ZebedeeClient) populateStaticLandingPageModel(dlp models.DatasetLandingPage) (StaticDatasetLandingPage, error) {
+	//Map Zebedee response data to new page model
+	var sdlp StaticDatasetLandingPage
+	sdlp.Type = dlp.Type
+	sdlp.URI = dlp.URI
+	sdlp.Metadata.Title = dlp.Description.Title
+	sdlp.Metadata.Description = dlp.Description.Summary
+	sdlp.DatasetLandingPage.Related.Datasets = dlp.RelatedDatasets
+	sdlp.DatasetLandingPage.Related.Publications = dlp.RelatedDocuments
+	sdlp.DatasetLandingPage.Related.Methodology = append(dlp.RelatedMethodology, dlp.RelatedMethodologyArticle...)
+	sdlp.DatasetLandingPage.Related.Links = dlp.RelatedLinks
+	sdlp.DatasetLandingPage.IsNationalStatistic = dlp.Description.NationalStatistic
+	sdlp.DatasetLandingPage.IsTimeseries = dlp.Timeseries
+	sdlp.ContactDetails.Email = dlp.Description.Contact.Email
```

Comment:
might be able to clean this up by wrapping some structs - if the source and dest struct have the same field names, Go 1.8+ can map it automatically, e.g. `sdlp.ContactDetails = Contact{dlp.Description.Contact}`